emulated/emulated.go

```diff
@@ -0,0 +1,366 @@
+		})
+	default:
+		r.logger.Debug("received unhandled app payload", "data", data)
+	}
+
+	return nil
+}
+
+func (r *Radio) nextPacketID() uint32 {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	r.packetID++
+	return r.packetID
+}
+
+func (r *Radio) sendPacket(ctx context.Context, packet *pb.MeshPacket) error {
+	// TODO: Optimistically attempt to encrypt the packet here if we recognise the channel, encryption is enabled and
+	// the payload is not currently encrypted.
+
+	// sendPacket is responsible for setting the packet ID.
+	r.packetID = r.nextPacketID()
+
+	se := &pb.ServiceEnvelope{
+		// TODO: Fetch channel to use based on packet.Channel rather than hardcoding to primary channel.
+		ChannelId: r.cfg.Channels.Settings[0].Name,
+		GatewayId: r.cfg.NodeID.String(),
+		Packet:    packet,
+	}
+	bytes, err := proto.Marshal(se)
+	if err != nil {
+		return fmt.Errorf("marshalling service envelope: %w", err)
+	}
+	return r.mqtt.Publish(&mqtt.Message{
+		Topic:   r.mqtt.GetFullTopicForChannel(r.cfg.Channels.Settings[0].Name) + "/" + r.cfg.NodeID.String(),
+		Payload: bytes,
+	})
+}
+
+func (r *Radio) broadcastNodeInfo(ctx context.Context) error {
+	r.logger.Info("broadcasting NodeInfo")
+	// TODO: Lots of stuff missing here. However, this is enough for it to show in the UI of another node listening to
+	// the MQTT server.
+	user := &pb.User{
+		Id:        r.cfg.NodeID.String(),
+		LongName:  r.cfg.LongName,
+		ShortName: r.cfg.ShortName,
+		HwModel:   pb.HardwareModel_PRIVATE_HW,
+	}
+	userBytes, err := proto.Marshal(user)
+	if err != nil {
+		return fmt.Errorf("marshalling user: %w", err)
+	}
+	return r.sendPacket(ctx, &pb.MeshPacket{
+		From: r.cfg.NodeID.Uint32(),
+		To:   meshtastic.BroadcastNodeID.Uint32(),
+		PayloadVariant: &pb.MeshPacket_Decoded{
+			Decoded: &pb.Data{
+				Portnum: pb.PortNum_NODEINFO_APP,
+				Payload: userBytes,
+			},
+		},
+	})
+}
+
+func (r *Radio) broadcastPosition(ctx context.Context) error {
+	r.logger.Info("broadcasting Position")
+	position := &pb.Position{
+		// TODO: Make broadcasting positional optional and configurable.
+		// Degrees divided by 1e-7 gives the value to use here.
+		// Buckingham Palace :D
+		LatitudeI:  515014760,
+		LongitudeI: -1406340,
+		Time:       uint32(time.Now().Unix()),
+	}
+	positionBytes, err := proto.Marshal(position)
+	if err != nil {
+		return fmt.Errorf("marshalling position: %w", err)
+	}
+	return r.sendPacket(ctx, &pb.MeshPacket{
+		From: r.cfg.NodeID.Uint32(),
+		To:   meshtastic.BroadcastNodeID.Uint32(),
+		PayloadVariant: &pb.MeshPacket_Decoded{
+			Decoded: &pb.Data{
+				Portnum: pb.PortNum_POSITION_APP,
+				Payload: positionBytes,
+			},
+		},
+	})
+}
+
+// dispatchMessageToFromRadio sends a FromRadio message to all current subscribers to
+// the FromRadio.
+func (r *Radio) dispatchMessageToFromRadio(msg *pb.FromRadio) error {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	for ch := range r.fromRadioSubscribers {
+		// TODO: Make this way safer/resilient
```

Comment:
There are a couple things you can do here:

1. Slow, but easy:  Just make these sends async with a timeout.
2. Faster, but more work:  Same, but in parallel across all the channels with a select to pick the ones that are ready and have a single timeout for the entire thing.

The second one gives you a total deadline for the entire delivery that will generally only be triggered after every handler that can receive has received.  But more work.

Reply:
Yeah we definitely need to do something here to prevent slow consumers from backing everything up. Since it's marked as a TODO already I'll tag this as resolved